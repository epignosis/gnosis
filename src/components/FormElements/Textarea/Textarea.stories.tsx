import React from "react";
import { Story } from "@storybook/react";
import TextareaComponent, { TextareaProps } from "./Textarea";

export default {
  title: "components/Form Elements/Textarea",
  component: TextareaComponent,
  args: {
    label: "Description",
    id: "textareaInput",
    inline: false,
    placeholder: "Type some description",
    resize: "none",
    status: "valid",
  },
  argTypes: {
    resize: {
      control: {
        type: "select",
        options: ["none", "vertical", "horizontal", "both"],
      },
    },
    status: {
      control: {
        type: "select",
        options: ["valid", "error"],
      },
    },
  },
  decorators: [
    (Story: Story): JSX.Element => (
      <div style={{ maxWidth: 500 }}>
        <Story />
      </div>
    ),
  ],
};

const Template: Story<TextareaProps> = (args) => <TextareaComponent {...args} />;

export const Default = Template.bind({});

export const Disabled = Template.bind({});

Disabled.args = {
  disabled: true,
};
