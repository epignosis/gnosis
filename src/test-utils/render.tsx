import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
