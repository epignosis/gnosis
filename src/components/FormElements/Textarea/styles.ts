import { css, SerializedStyles, Theme } from "@emotion/react";
import { inputBaseStyles, inputContainerBaseStyles } from "../styles";
import { TextareaResize } from "./Textarea";

export const textareaContainer = (
  { formElements }: Theme,
  { resize }: { resize: TextareaResize },
): SerializedStyles => css`
  ${inputContainerBaseStyles({ block: true })};

  &.inline {
    align-items: baseline;
  }

  &.error {
    textarea {
      border-color: ${formElements.errors.errorBorderColor};
    }
  }

  textarea {
    ${inputBaseStyles(formElements, { block: true })};
    min-height: 10rem;
    padding: 0.75rem;
    resize: ${resize};
  }
`;
