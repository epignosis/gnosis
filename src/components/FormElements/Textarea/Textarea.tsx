import React, { forwardRef, ForwardRefRenderFunction, InputHTMLAttributes } from "react";
import { SerializedStyles } from "@emotion/react";
import classNames from "classnames";
import Label from "../Label/Label";
import { textareaContainer } from "./styles";
import { ExtendableProps } from "types/common";

export type TextareaResize = "none" | "horizontal" | "vertical" | "both";

export type TextareaProps = ExtendableProps<
  InputHTMLAttributes<HTMLTextAreaElement>,
  {
    id: string;
    label?: string;
    inline?: boolean;
    resize?: TextareaResize;
    status?: "valid" | "error";
    containerAttrs?: React.HTMLAttributes<HTMLDivElement>;
  }
>;
const Textarea: ForwardRefRenderFunction<HTMLTextAreaElement, TextareaProps> = (
  props,
  forwardedRef,
) => {
  const {
    label,
    inline = false,
    id,
    resize = "none",
    status = "valid",
    containerAttrs,
    ...rest
  } = props;
  const hasLabel = Boolean(label);
  const containerClassNames = classNames({
    inline: hasLabel && inline,
    valid: status === "valid",
    error: status === "error",
    [containerAttrs?.className ?? ""]: Boolean(containerAttrs?.className),
  });

  return (
    <div
      css={(theme): SerializedStyles => textareaContainer(theme, { resize })}
      {...containerAttrs}
      className={containerClassNames}
    >
      {hasLabel && <Label htmlFor={id}>{label}</Label>}
      <textarea id={id} ref={forwardedRef} {...rest} />
    </div>
  );
};

export default forwardRef(Textarea);
