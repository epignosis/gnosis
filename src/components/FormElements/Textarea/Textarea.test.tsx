import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Textarea from "./Textarea";
import { screen, render } from "@test-utils/render";

describe("<Textarea />", () => {
  it("renders correctly", () => {
    const labelTxt = faker.random.words();
    render(
      <Textarea
        id={faker.random.alphaNumeric()}
        label={labelTxt}
        name={faker.random.alphaNumeric()}
      />,
    );

    const textarea = screen.getByLabelText(labelTxt);

    expect(textarea).toBeInTheDocument();
  });

  it("changes value", () => {
    const labelTxt = faker.random.words();
    render(
      <Textarea
        id={faker.random.alphaNumeric()}
        label={labelTxt}
        name={faker.random.alphaNumeric()}
      />,
    );

    const textarea = screen.getByLabelText(labelTxt);

    expect(textarea).toHaveValue("");

    const newValue = faker.lorem.paragraph();
    userEvent.type(textarea, newValue);

    expect(textarea).toHaveValue(newValue);
  });

  it("matches snapshot", () => {
    const { container } = render(<Textarea id="test-textarea" label="Test textarea" name="test" />);

    expect(container).toMatchSnapshot();
  });
});
