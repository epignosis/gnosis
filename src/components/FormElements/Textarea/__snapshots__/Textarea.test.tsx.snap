// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea /> matches snapshot 1`] = `
<div>
  <div
    class="valid textarea-container-class css-11n4xhq-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-textareaContainer-textareaContainer"
    id="textarea-container-id"
  >
    <label
      class="css-xs6o3g-label-label"
      for="my-textarea-id"
    >
      Test textarea
    </label>
    <textarea
      class="textarea-class"
      id="my-textarea-id"
      name="test"
    />
  </div>
</div>
`;
